common/list.c

```diff
@@ -298,3 +297,40 @@ list_dump_items(struct list *self)
         LOG_DEVEL(LOG_LEVEL_TRACE, "%d: %p", index, (void *) list_get_item(self, index));
     }
 }
+
+/******************************************************************************/
+static const char *
+append_fragment(const char *start, const char *end, struct list *list)
+{
+    const unsigned int len = end - start;
+    char *copy = (char *)malloc(len + 1);
+    g_memcpy(copy, start, len);
+    copy[len] = '\0';
+    list_add_item(list, (tintptr)copy);
+    return end + 1;
+}
+
+/******************************************************************************/
+struct list *
+split_string_into_list(const char *str, char character)
+{
+    struct list *result = list_create();
+    result->auto_free = 1;
+
+    if (str == NULL)
+    {
+        return result;
+    }
+
+    const char *p;
+    while ((p = g_strchr(str, character)) != NULL)
+    {
+        str = append_fragment(str, p, result);
+    }
+
+    if (*str != '\0')
+    {
+        append_fragment(str, str + g_strlen(str), result);
+    }
+    return result;
+}
```

Comment:
Sorry to be a bit of a pain, but after #2536 was merged last week we can check for allocation failures, so we probably ought to for new functions like this.

Here's a modified version. Feel free to hack it about as appropriate:-

```c
/******************************************************************************/
/**
 * Appends a string fragment to a list
 * @param[in,out] start Pointer to start of fragment (by reference)
 * @param end Pointer to one past end of fragment
 * @param list List to append to
 * @result 1 for success
 *
 * In the event of a memory failure, 0 is returned and the list is deleted
 */
static int
append_fragment(const char **start, const char *end, struct list *list)
{
    const unsigned int len = end - *start;
    char *copy = (char *)malloc(len + 1);
    if (copy == NULL)
    {
        list_delete(list);
        return 0;
    }
    g_memcpy(copy, *start, len);
    copy[len] = '\0';
    if (!list_add_item(list, (tintptr)copy))
    {
        g_free(copy);
        list_delete(list);
        return 0;
    }
    *start = end + 1;
    return 1;
}

/******************************************************************************/
struct list *
split_string_into_list(const char *str, char character)
{
    struct list *result = list_create();
    if (result == NULL)
    {
        return result;
    }
    result->auto_free = 1;

    if (str == NULL)
    {
        return result;
    }

    const char *p;
    while ((p = g_strchr(str, character)) != NULL)
    {
        if (!append_fragment(&str, p, result))
        {
            return NULL;
        }
    }

    if (*str != '\0')
    {
        if (!append_fragment(&str, str + g_strlen(str), result))
        {
            return NULL;
        }
    }
    return result;
}
```

Reply:
Not a pain at all! Working on updating